staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -496,6 +509,32 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 		chanSize = 1000
 	}
 
+	// We explicitly use thread unsafe version and do locking ourself to ensure that
+	// no new events will be processed in the meantime. The watchCache will be unlocked
+	// on return from this function.
+	// Note that we cannot do it under Cacher lock, to avoid a deadlock, since the
+	// underlying watchCache is calling processEvent under its lock.
+	c.watchCache.RLock()
+	defer c.watchCache.RUnlock()
```

Comment:
Why do we need to move the lock here?

IIUC, the only point where we really need that is when `watchRV==0` (to the the `watchCache.resourceVersion`).

If that's correct, then I would like to do that differently. There was a reason why the locking is happening relatively late - basically, to avoid memory allocations under lock (which historically was visibly increasing lock contention).
So what I would do basically is:
- revert the locking to where it was
- stop passing `bookmarkAfterResourceVersion` parameter as a newCacheWatcher parameter
- instead, set the bookmarkAfterResourceVersion field explicitly in the same place where its forget field is set

Reply:
Modulo the fact that we should never be calling etcd under lock - that will be  disaster.

So what we need is basically something like that:
```
func (c *Cacher) getBookmarkAfterResourceVersionFunc(opts) func() uint64 {
  if !opts.SendInitialEvents && utilfeature... {
    return func() uint64 { return 0 }
  }
  switch {
  case len(opts.ResourceVersion) == 0:
    rv, err := c.getCurrentResourceVersionFromStorage(ctx)
    if err != nil {...}
    return func() uint64 { return rv }
  case watchRV == 0:
    // Here we assume that watchCache locked is already held.
    return func() uint64 { return c.watchCache.resourceVersion }
  default:
    return func() uint64 { return watchRV }
  }
}
```

Reply:
yes, I moved the lock to get `watchCache.resourceVersion` for `watchRV==0`

> There was a reason why the locking is happening relatively late - basically, to avoid memory allocations under lock

are you referring to calling `newCacheWatcher` ? I think this is the only thing that has changed after moving the lock.

> Modulo the fact that we should never be calling etcd under lock - that will be disaster.

Could you please explain what it would be a disaster?